libp2p/NodeTable.cpp

```diff
@@ -328,50 +315,46 @@ void NodeTable::noteActiveNode(Public const& _pubk, bi::udp::endpoint const& _en
     if (_pubk == m_node.address() || !NodeIPEndpoint(_endpoint.address(), _endpoint.port(), _endpoint.port()).isAllowed())
         return;
 
-    shared_ptr<NodeEntry> node = nodeEntry(_pubk);
-    if (!!node && !node->pending)
+    shared_ptr<NodeEntry> newNode = nodeEntry(_pubk);
+    if (newNode && !newNode->pending)
     {
         clog(NodeTableConnect) << "Noting active node:" << _pubk << _endpoint.address().to_string() << ":" << _endpoint.port();
-        node->endpoint.address = _endpoint.address();
-        node->endpoint.udpPort = _endpoint.port();
-        
-        shared_ptr<NodeEntry> contested;
+        newNode->endpoint.address = _endpoint.address();
+        newNode->endpoint.udpPort = _endpoint.port();
+
+        shared_ptr<NodeEntry> nodeToEvict;
         {
+            // Find a bucket to put a node to
             Guard l(x_state);
-            NodeBucket& s = bucket_UNSAFE(node.get());
-            bool removed = false;
-            s.nodes.remove_if([&node, &removed](weak_ptr<NodeEntry> const& n)
-            {
-                if (n.lock() == node)
-                    removed = true;
-                return removed;
-            });
-            
-            if (s.nodes.size() >= s_bucketSize)
+            NodeBucket& s = bucket_UNSAFE(newNode.get());
+
+            if (s.nodes.size() < s_bucketSize)
             {
-                if (removed)
-                    clog(NodeTableWarn) << "DANGER: Bucket overflow when swapping node position.";
-                
-                // It's only contested iff nodeentry exists
-                contested = s.nodes.front().lock();
-                if (!contested)
```

Comment:
This condition was always `false` I think, because all `shared_ptr<NodeEntry>` are never destroyed, so `lock()` above was always successful. 
Sp I remove this part

Reply:
Writing the tests I found that I was not right here, `shared_ptr<NodeEntry>` can be destroyed when overwritten with `m_nodes[_node.id] = ret;` in `NodeTable::addNode()`

Reply:
Yes, I've brought it back now https://github.com/ethereum/cpp-ethereum/pull/4920/files#diff-b79ec8f203ebd2b76e119c5c3bc8b242R355
`contested` is just renamed to `nodeToEvict`